rest-service/manager_rest/security/user_handler.py

```diff
@@ -1,67 +1,159 @@
 from datetime import datetime
 import string
+import typing
 
-from flask import current_app
+from flask import current_app, Response, abort, Request
 from flask_security.utils import verify_password
 
+from cloudify.workflows import tasks
+from cloudify.models_states import ExecutionState
+
 from manager_rest.storage.models import Token, User
 from manager_rest.storage.models_base import db
 from manager_rest.manager_exceptions import (
     NoAuthProvided,
     UnauthorizedError,
 )
 from manager_rest.storage import user_datastore, get_storage_manager
-from manager_rest.execution_token import (set_current_execution,
-                                          get_current_execution_by_token,
-                                          get_execution_token_from_request)
-from manager_rest.utils import is_expired
+from manager_rest.execution_token import (
+    set_current_execution,
+    get_current_execution_by_token,
+    get_execution_token_from_request,
+    current_execution,
+)
+from manager_rest.security import audit
+from manager_rest.utils import (
+    is_expired,
+    is_sanity_mode,
+)
 
 
-ENCODED_ID_LENGTH = 5
+def user_loader(request: Request) -> typing.Optional[User]:
+    """Load a user object based on the request.
 
+    This function is passed as a flask-login request_loader, and will be
+    called for all requests.
 
-def user_loader(request):
-    """Attempt to retrieve the current user from the request
-    Either from request's Authorization attribute, or from the token header
+    Try several ways of retrieving the user: each of these retrieval functions
+    are expected to decide whether they should be the one to get the user,
+    and return one, or None. The functions can also throw, or return an error
+    response immediately.
+    """
+    for getter, method_label in [
+        (_get_user_from_execution_token, 'execution_token'),
+        (_get_user_from_auth, 'http_basic'),
```

Comment:
Doesn't this ordering result in ldap breaking as it uses http basic auth then passes through?

Reply:
Good point. The `7da7b126b Only allow the bootstrap admin to bypass ext auth` commit should address this, and keep the previous behaviour. Now, if ext-auth is configured, this function will be a no-op for all but the bootstrap admin.